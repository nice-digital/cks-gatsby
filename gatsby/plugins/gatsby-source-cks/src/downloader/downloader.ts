/**
 * Downloads all data (topics, their associated data and the latest change) from
 * the API, reporting progress and status to gatsby.
 */

import { SourceNodesArgs } from "gatsby";
import { mapLimit, asyncify } from "async";

import {
	getAllPartialTopics,
	getChangesSince,
	getFullTopicCached,
} from "../api/api";

import {
	ApiTopicChangeResponse,
	ApiPartialTopic,
	ApiFullTopic,
	ApiTopicsResponse,
} from "../api/types";

interface DownloadResult {
	fullTopics: ApiFullTopic[];
	changes: ApiTopicChangeResponse;
}

const getAllFullTopics = async (
	{ topicCount, topics: partialTopics }: ApiTopicsResponse,
	{ cache, reporter: { createProgress } }: SourceNodesArgs
): Promise<ApiFullTopic[]> => {
	const topicDownloadProgress = createProgress(
		`Downloading ${topicCount} topics`,
		topicCount,
		0
	);
	topicDownloadProgress.start();

	// Load each full topic async in parallel but rate limited:
	// The API can't cope with firing them all off in parallel
	const fullTopics = await mapLimit<ApiPartialTopic, ApiFullTopic>(
		partialTopics,
		5,
		// Because we're transpiling an async function we need 'asyncify' it
		// see https://caolan.github.io/async/v3/docs.html#asyncify
		asyncify((partialTopic: ApiPartialTopic) => {
			const fullTopic = getFullTopicCached(partialTopic, cache);
			topicDownloadProgress.tick(1);
			return fullTopic;
		})
	);

	topicDownloadProgress.done();

	return fullTopics;
};

export const downloadAllData = async (
	sourceNodesArgs: SourceNodesArgs,
	changesSinceDate: Date
): Promise<DownloadResult> => {
	const {
		reporter: { activityTimer, info },
	} = sourceNodesArgs;

	const apiActivity = activityTimer(`Downloading data from the API`);
	apiActivity.start();

	apiActivity.setStatus(`Downloading topics index`);
	const partialTopicsTask = getAllPartialTopics();

	apiActivity.setStatus(`Downloading monthly changes`);
	const whatsNewChangesTask = getChangesSince(changesSinceDate);

	const changes = await whatsNewChangesTask,
		fullTopics = await getAllFullTopics(
			await partialTopicsTask,
			sourceNodesArgs
		);

	changes.forEach(c => {
		info(`Got major update for '${c.topicName}' (${c.title})`);
	});

	apiActivity.setStatus(`Received all data from the API`);
	apiActivity.end();

	return {
		fullTopics,
		changes,
	};
};
