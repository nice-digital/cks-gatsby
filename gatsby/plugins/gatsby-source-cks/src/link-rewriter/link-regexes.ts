/**
 * Anchors to a different topic.
 *
 * @example
 * <a class="topic-reference external-reference" href="/Topic/ViewTopic/a549c958-335f-4dcd-b76d-e9f87325d888">Chest pain</a>
 */
export const topicAnchorRegex = /<a(?:(?!<a).)*href="(\/Topic\/ViewTopic\/(.{36}))".*?<\/a>/gi;

/**
 * Anchors to a chapter within the same topic.
 *
 * @example
 * <a class="bibliography-reference internal-reference" href="#666cef38-2cdb-45b8-a427-aaffb3caa682">NICE, 2016a</a>
 */
export const chapterAnchorRegex = /<a(?:(?!<a).)*href="(#(.{36}))".*?<\/a>/gi;

/**
 * Anchors to a chapter within a different topic.
 *
 * @example
 * <a
 * 	class="topic-reference external-reference"
 * 	href="/Topic/ViewTopic/a549c958-335f-4dcd-b76d-e9f87325d888#67439879-2ad5-4721-b865-5aadd9bebe52">
 * 		chest pain
 * </a>
 */
export const topicChapterAnchorRegex = /<a(?:(?!<a).)*href="(\/Topic\/ViewTopic\/(.{36})#(.{36}))".*?<\/a>/gi;
