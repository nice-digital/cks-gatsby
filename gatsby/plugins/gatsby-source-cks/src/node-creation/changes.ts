/**
 * Creates Gatsby GraphQL nodes for monthly changes (for the what's new page) from downloaded data
 */

import { NodeInput, SourceNodesArgs } from "gatsby";
import { ApiTopicChangeResponse } from "../api/types";

export const changeNodeType = "CksChange";

export interface ChangeNode extends NodeInput {
	title: string;
	text: string;
	// Foreign keys
	topic: string;
	internal: {
		type: typeof changeNodeType;
	} & NodeInput["internal"];
}

export const createChangeNodes = (
	changes: ApiTopicChangeResponse,
	sourceNodesArgs: SourceNodesArgs
): void => {
	const { createNodeId, createContentDigest, actions } = sourceNodesArgs;
	const { createNode } = actions;

	changes.forEach(({ topicId, title, text }) => {
		const nodeContent = {
			title,
			text,
			topic: topicId,
		};

		const changeNode: ChangeNode = {
			...nodeContent,
			id: createNodeId(`Change${topicId}`),
			children: [],
			internal: {
				type: changeNodeType,
				content: JSON.stringify(nodeContent),
				contentDigest: createContentDigest(nodeContent),
			},
		};

		createNode(changeNode);
	});
};
