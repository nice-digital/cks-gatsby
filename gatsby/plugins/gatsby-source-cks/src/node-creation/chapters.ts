/**
 * Creates Gatsby GraphQL nodes for chapters (HTML objects within topics) from downloaded data
 */

import { NodeInput, SourceNodesArgs } from "gatsby";
import slugify from "slugify";

import { ApiFullTopic, ApiTopicHtmlObject } from "../api/types";

const BasisChapterTitle = "Basis for recommendation";

export const chapterNodeType = "CksChapter";

export interface ChapterNode extends NodeInput {
	slug: string;
	itemId: string;
	fullItemName: string;
	htmlHeader: string;
	htmlStringContent: string;
	containerElement: string;
	depth: number;
	pos: number;
	// Foreign keys
	topic: string;
	parentChapter: string | null;
	rootChapter: string;
	subChapters: string[];
	children?: undefined;
	internal: {
		type: typeof chapterNodeType;
	} & NodeInput["internal"];
}

const createTopicChapterNodes = (
	topicId: string,
	topicHtmlObjects: ApiTopicHtmlObject[],
	sourceNodesArgs: SourceNodesArgs
): void => {
	const { createNodeId, createContentDigest, actions } = sourceNodesArgs;
	const { createNode } = actions;

	topicHtmlObjects.forEach(
		({
			itemId,
			containerElement,
			parentId,
			rootId,
			children,
			fullItemName,
			...chapterFields
		}) => {
			let slug = slugify(
				fullItemName.replace(/^Scenario: /gi, "").replace(/ and /gi, " "),
				{
					lower: true,
					remove: /[,*+~.()'"!?:@]/g,
				}
			);

			// There can be multiple basis for recs on the same page, so distinguish the slugs
			// See example at: /topics/angina/management/new-diagnosis/
			if (fullItemName === BasisChapterTitle) {
				slug = `${slug}-${itemId.substring(0, 3)}`;
			}

			const nodeContent = {
				...chapterFields,
				slug,
				fullItemName,
				containerElement,
				itemId,
				topic: topicId,
				parentChapter: parentId,
				rootChapter: rootId,
				subChapters: children.map((c) => c.itemId),
			};

			const chapterNode: ChapterNode = {
				...nodeContent,
				id: createNodeId(itemId),
				internal: {
					type: chapterNodeType,
					content: JSON.stringify(nodeContent),
					contentDigest: createContentDigest(nodeContent),
				},
			};

			createNode(chapterNode);

			// Recursively create sub chapters
			createTopicChapterNodes(topicId, children, sourceNodesArgs);
		}
	);
};

export const createChapterNotes = (
	topics: ApiFullTopic[],
	sourceNodesArgs: SourceNodesArgs
): void => {
	topics.forEach(({ topicId, topicHtmlObjects }) => {
		createTopicChapterNodes(topicId, topicHtmlObjects, sourceNodesArgs);
	});
};
