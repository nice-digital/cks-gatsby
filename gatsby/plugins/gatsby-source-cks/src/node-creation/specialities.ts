/**
 * Creates Gatsby GraphQL nodes for specialities (that group topics together) from downloaded data
 */

import { SourceNodesArgs, NodeInput } from "gatsby";
import slugify from "slugify";

import { ApiFullTopic } from "../api/types";

const specialityNodeType = "CksSpeciality";

interface SpecialityNode extends NodeInput {
	name: string;
	slug: string;
	// Foreign keys
	topics: string[];
	internal: {
		type: typeof specialityNodeType;
	} & NodeInput["internal"];
}

/**
 * Pull out each of the topics's clinicalSpecialties into
 * a mapping of speciality name to topic ids
 * e.g. { 'Skin and nail': [ 1, 2 ], Cardiovascular: [ 1, 3 ], ... }
 *
 * @param topics
 */
const extractSpecialities = (topics: ApiFullTopic[]): Map<string, string[]> => {
	const mapping: Map<string, string[]> = new Map();

	topics.forEach(({ topicId, clinicalSpecialties }) => {
		clinicalSpecialties.forEach((speciality) =>
			mapping.set(speciality, [...(mapping.get(speciality) || []), topicId])
		);
	});

	return mapping;
};

export const createSpecialityNodes = (
	fullTopics: ApiFullTopic[],
	sourceNodesArgs: SourceNodesArgs
): void => {
	const { createNodeId, createContentDigest, actions } = sourceNodesArgs;
	const { createNode } = actions;

	const specialityTopicsMapping = extractSpecialities(fullTopics);

	Array.from(specialityTopicsMapping.entries()).forEach(([name, topics]) => {
		const nodeContent = {
			name,
			slug: slugify(name, { lower: true, remove: /[(),']/g }),
			topics,
		};

		const specialityNode: SpecialityNode = {
			...nodeContent,
			id: createNodeId(name),
			internal: {
				type: specialityNodeType,
				content: JSON.stringify(nodeContent),
				contentDigest: createContentDigest(nodeContent),
			},
		};

		createNode(specialityNode);
	});
};
