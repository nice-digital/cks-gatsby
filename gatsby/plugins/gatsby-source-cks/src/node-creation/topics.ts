/**
 * Creates Gatsby GraphQL nodes for topics from downloaded data
 */

import { SourceNodesArgs, NodeInput } from "gatsby";
import slugify from "slugify";

import { ApiSingleTopicResponse, ApiFullTopic } from "../api/types";

export const topicNodeType = "CksTopic";

export interface TopicNode
	extends NodeInput,
		Omit<ApiSingleTopicResponse, "clinicalSpecialties" | "topicHtmlObjects"> {
	slug: string;
	// Foreign keys
	specialities: string[];
	chapters: string[];
	// Internal Gatsby node stuff
	internal: {
		type: typeof topicNodeType;
	} & NodeInput["internal"];
}

export const createTopicNodes = (
	topics: ApiFullTopic[],
	sourceNodesArgs: SourceNodesArgs
): void => {
	const { createNodeId, createContentDigest, actions } = sourceNodesArgs;
	const { createNode } = actions;

	topics.forEach(
		({ topicHtmlObjects, clinicalSpecialties, ...topic }: ApiFullTopic) => {
			const { topicId, topicName } = topic;

			const nodeContent = {
				...topic,
				...{
					slug: slugify(topicName, { lower: true, remove: /[(),']/g }),
					specialities: clinicalSpecialties,
					chapters: topicHtmlObjects.map((t) => t.itemId),
				},
			};

			const topicNode: TopicNode = {
				...nodeContent,
				id: createNodeId(topicId),
				children: [],
				internal: {
					type: topicNodeType,
					content: JSON.stringify(nodeContent),
					contentDigest: createContentDigest(nodeContent),
				},
			};

			createNode(topicNode);
		}
	);
};
