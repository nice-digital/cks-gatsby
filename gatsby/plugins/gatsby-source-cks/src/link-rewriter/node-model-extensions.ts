import { NodeInput } from "gatsby";

import { NodeModel } from "./types";
import { TopicNode, topicNodeType } from "../node-creation/topics";
import { ChapterNode, chapterNodeType } from "../node-creation/chapters";

const TopicIdField = "topicId";
const ChapterIdField = "itemId";

interface GetNodeByIdArgs {
	idField: string;
	id: string;
	nodeTypeName: string;
	nodeModel: NodeModel;
}

const getNodeById = <T extends NodeInput>({
	idField,
	id,
	nodeTypeName,
	nodeModel,
}: GetNodeByIdArgs): Promise<T | null> =>
	nodeModel.findOne({
		query: {
			filter: {
				[idField]: { eq: id },
			},
		},
		type: nodeTypeName,
	});

export const getTopicById = (
	topicId: string,
	nodeModel: NodeModel
): Promise<TopicNode | null> =>
	getNodeById({
		idField: TopicIdField,
		id: topicId,
		nodeTypeName: topicNodeType,
		nodeModel,
	});

export const getChapterById = (
	chapterItemId: string,
	nodeModel: NodeModel
): Promise<ChapterNode | null> =>
	getNodeById({
		idField: ChapterIdField,
		id: chapterItemId,
		nodeTypeName: chapterNodeType,
		nodeModel,
	});
