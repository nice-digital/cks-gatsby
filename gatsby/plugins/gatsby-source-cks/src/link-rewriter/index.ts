import { Reporter } from "gatsby";
import { NodeModel } from "./types";
import { ChapterNode } from "../node-creation/chapters";
import { replaceAsync } from "../utils";

import { getTopicById, getChapterById } from "./node-model-extensions";
import {
	topicAnchorRegex,
	chapterAnchorRegex,
	topicChapterAnchorRegex,
} from "./link-regexes";

interface LinkRewriteArgs {
	htmlStringContent: string;
	nodeModel: NodeModel;
	//reporter: Reporter;
	//getErrorMessage: (message: string, anchor: string) => string;
	warn: (message: string, anchor: string) => void;
	error: (message: string, anchor: string) => void;
}

/**
 * Topic links are in the form /Topic/ViewTopic/<topic guid> in the feed.
 * So these need to be written into the form /topics/<slug>
 */
const rewriteTopicLinks = ({
	htmlStringContent,
	nodeModel,
	warn,
}: LinkRewriteArgs): Promise<string> => {
	return replaceAsync(
		htmlStringContent,
		topicAnchorRegex,
		async (anchor: string, originalHref: string, targetTopicId: string) => {
			const targetTopic = await getTopicById(targetTopicId, nodeModel);

			if (!targetTopic) {
				warn(`Could not find topic '${targetTopicId}'`, anchor);
				return anchor;
			}

			return anchor.replace(originalHref, `/topics/${targetTopic.slug}/`);
		}
	);
};

const rewriteChapterLinks = ({
	htmlStringContent,
	nodeModel,
	warn,
	error,
}: LinkRewriteArgs): Promise<string> => {
	return replaceAsync(
		htmlStringContent,
		chapterAnchorRegex,
		async (anchor: string, originalHref: string, targetChapterId: string) => {
			const targetChapter = await getChapterById(targetChapterId, nodeModel);

			if (!targetChapter) {
				warn(`Could not find chapter '${targetChapterId}'`, anchor);
				return anchor;
			}

			const targetChapterhref = await getChapterHref(targetChapter, nodeModel);

			if (!targetChapterhref) {
				error(`6th level link found in '${targetChapterId}'`, anchor);
				return anchor;
			}

			return anchor.replace(originalHref, targetChapterhref);
		}
	);
};

const rewriteTopicChapterLinks = ({
	htmlStringContent,
	nodeModel,
	warn,
	error,
}: LinkRewriteArgs): Promise<string> =>
	replaceAsync(
		htmlStringContent,
		topicChapterAnchorRegex,
		async (
			anchor: string,
			originalHref: string,
			targetTopicId: string,
			targetChapterId: string
		) => {
			const targetTopic = await getTopicById(targetTopicId, nodeModel);

			if (!targetTopic) {
				warn(`Could not find topic '${targetTopicId}'`, anchor);
				return anchor;
			}

			const targetChapter = await getChapterById(targetChapterId, nodeModel);

			if (!targetChapter) {
				warn(`Could not find chapter '${targetChapterId}'`, anchor);
				return anchor;
			} else if (targetChapter.topic !== targetTopicId) {
				warn(
					`Chapter '${targetChapterId}' topic (${targetChapter.topic}) doesn't match topic in href (${targetTopicId})`,
					anchor
				);
				return anchor;
			}

			const targetChapterhref = await getChapterHref(targetChapter, nodeModel);

			if (!targetChapterhref) {
				error(`6th level link found in '${targetChapterId}'`, anchor);
				return anchor;
			}

			return anchor.replace(originalHref, targetChapterhref);
		}
	);

const getChapterHref = async (
	targetChapterNode: ChapterNode,
	nodeModel: NodeModel
): Promise<string | null> => {
	const topicNode = await getTopicById(targetChapterNode.topic, nodeModel),
		topicPath = `/topics/${topicNode?.slug}/`;

	// Top level chapter
	if (!targetChapterNode.parentChapter) {
		return `${topicPath}${targetChapterNode.slug}/`;
	}

	// Second level chapter
	const parentChapterNode = await getChapterById(
		targetChapterNode.parentChapter,
		nodeModel
	);
	if (parentChapterNode?.itemId === targetChapterNode.rootChapter) {
		return `${topicPath}${parentChapterNode?.slug}/${targetChapterNode.slug}/`;
	}

	// Third level chapter
	const grandParentChapterNode = await getChapterById(
		parentChapterNode?.parentChapter as string,
		nodeModel
	);
	if (grandParentChapterNode?.itemId === targetChapterNode.rootChapter) {
		return `${topicPath}${grandParentChapterNode?.slug}/${parentChapterNode?.slug}/#${targetChapterNode.slug}`;
	}

	// Fourth level chapter
	const greatGrandParentChapterNode = await getChapterById(
		grandParentChapterNode?.parentChapter as string,
		nodeModel
	);
	if (greatGrandParentChapterNode?.itemId === targetChapterNode.rootChapter) {
		return `${topicPath}${greatGrandParentChapterNode?.slug}/${grandParentChapterNode?.slug}/#${targetChapterNode.slug}`;
	}

	// Fifth level chapter
	const greatGreatGrandParentChapterNode = await getChapterById(
		greatGrandParentChapterNode?.parentChapter as string,
		nodeModel
	);
	if (
		greatGreatGrandParentChapterNode?.itemId === targetChapterNode.rootChapter
	) {
		return `${topicPath}${greatGreatGrandParentChapterNode?.slug}/${greatGrandParentChapterNode?.slug}/#${targetChapterNode.slug}`;
	}

	// 6th level isn't supported
	return null;
};

export const replaceLinksInHtml = async (
	currentChapter: ChapterNode,
	nodeModel: NodeModel,
	reporter: Reporter
): Promise<string> => {
	const { htmlStringContent } = currentChapter;
	const { topic: topicId } = currentChapter;

	const currentTopic = await getTopicById(topicId, nodeModel);

	if (!currentTopic) {
		reporter.error(`Could not find topic with id ${topicId}`);
		return "Error";
	}

	// Generates a useful error message with the anchor location to make it easier to find the bad link in the feed
	const getErrorMessage = (message: string, anchor: string) =>
		[
			message,
			` in link ${anchor}`,
			` in chapter '${currentChapter.fullItemName}' (${currentChapter.itemId})`,
			` in topic '${currentTopic.topicName}' (${currentTopic.topicId})`,
		].join("\n");

	const args: LinkRewriteArgs = {
		htmlStringContent,
		nodeModel,
		warn: (message: string, anchor: string) =>
			reporter.warn(getErrorMessage(message, anchor)),
		error: (message: string, anchor: string) =>
			reporter.error(getErrorMessage(message, anchor)),
	};

	args.htmlStringContent = await rewriteTopicLinks(args);
	args.htmlStringContent = await rewriteChapterLinks(args);
	args.htmlStringContent = await rewriteTopicChapterLinks(args);

	return args.htmlStringContent;
};
