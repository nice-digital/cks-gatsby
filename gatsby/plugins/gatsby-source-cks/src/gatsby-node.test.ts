import {
	createSchemaCustomization,
	sourceNodes,
	createResolvers,
} from "./gatsby-node";
import {
	CreateSchemaCustomizationArgs,
	SourceNodesArgs,
	CreateResolversArgs,
} from "gatsby";

import { downloadAllData } from "./downloader/downloader";
import { createTopicNodes } from "./node-creation/topics";
import { createSpecialityNodes } from "./node-creation/specialities";
import { createChangeNodes } from "./node-creation/changes";
import { createChapterNotes } from "./node-creation/chapters";
import { replaceLinksInHtml } from "./link-rewriter";

jest
	.mock("./downloader/downloader", () => ({
		downloadAllData: jest.fn().mockResolvedValue({}),
	}))
	.mock("./node-creation/topics", () => ({
		createTopicNodes: jest.fn(),
	}))
	.mock("./node-creation/specialities", () => ({
		createSpecialityNodes: jest.fn(),
	}))
	.mock("./node-creation/changes", () => ({
		createChangeNodes: jest.fn(),
	}))
	.mock("./node-creation/chapters", () => ({
		createChapterNotes: jest.fn(),
	}))
	.mock("./link-rewriter", () => ({
		replaceLinksInHtml: jest.fn(),
	}));

describe("gatsby-node", () => {
	describe("createSchemaCustomization", () => {
		it("should call createTypes with custom schema", () => {
			const createTypes = jest.fn();

			const args: unknown = {
				actions: { createTypes },
			};

			createSchemaCustomization(args as CreateSchemaCustomizationArgs);

			expect(createTypes).toHaveBeenCalledTimes(1);

			expect(createTypes.mock.calls[0][0]).toMatchSnapshot();
		});
	});

	describe("sourceNodes", () => {
		const sourceNodesArgs = ({
			reporter: {
				activityTimer: (): unknown => ({
					start: (): void => void 0,
					end: (): void => void 0,
					setStatus: (): void => void 0,
				}),
				info: (): void => void 0,
			},
		} as unknown) as SourceNodesArgs;

		const config = {
			apiBaseUrl: "t",
			apiKey: "b",
			changesSinceDate: new Date(2020, 0, 12),
		};

		beforeEach(() => {
			jest.clearAllMocks();
		});

		it("should download all data", async () => {
			await sourceNodes(sourceNodesArgs, config);

			expect(downloadAllData).toHaveBeenCalledTimes(1);
			expect(downloadAllData).toHaveBeenCalledWith(
				sourceNodesArgs,
				config.changesSinceDate
			);
		});

		it("should create all 4 types of nodes", async () => {
			(downloadAllData as jest.Mock).mockResolvedValueOnce({
				fullTopics: [{ a: 1 }],
				changes: [{ b: 2 }],
			});

			await sourceNodes(sourceNodesArgs, config);

			expect(createTopicNodes).toHaveBeenCalledTimes(1);
			expect(createTopicNodes).toHaveBeenCalledWith(
				[{ a: 1 }],
				sourceNodesArgs
			);
			expect(createSpecialityNodes).toHaveBeenCalledTimes(1);
			expect(createSpecialityNodes).toHaveBeenCalledWith(
				[{ a: 1 }],
				sourceNodesArgs
			);
			expect(createChangeNodes).toHaveBeenCalledTimes(1);
			expect(createChangeNodes).toHaveBeenCalledWith(
				[{ b: 2 }],
				sourceNodesArgs
			);
			expect(createChapterNotes).toHaveBeenCalledTimes(1);
			expect(createChapterNotes).toHaveBeenCalledWith(
				[{ a: 1 }],
				sourceNodesArgs
			);
		});
	});

	describe("createResolvers", () => {
		it("should create a resolver for Chapter htmlStringContent property", () => {
			const createResolversFn = jest.fn();

			createResolvers({
				createResolvers: createResolversFn as unknown,
			} as CreateResolversArgs);

			expect(createResolversFn).toHaveBeenCalledTimes(1);
			expect(createResolversFn).toHaveBeenCalledWith({
				CksChapter: { htmlStringContent: { resolve: expect.any(Function) } },
			});
		});

		it("should create resolver that calls replaceLinksInHtml with chapter and node model", async () => {
			const createResolversFn = jest.fn();

			createResolvers({
				createResolvers: createResolversFn as unknown,
			} as CreateResolversArgs);

			const chapter = { slug: "test" };
			const resolveContext = { nodeModel: { runQuery: jest.fn() } };

			await createResolversFn.mock.calls[0][0].CksChapter.htmlStringContent.resolve(
				chapter,
				{},
				resolveContext
			);

			expect(replaceLinksInHtml).toHaveBeenCalledWith(
				chapter,
				resolveContext.nodeModel
			);
		});
	});
});
