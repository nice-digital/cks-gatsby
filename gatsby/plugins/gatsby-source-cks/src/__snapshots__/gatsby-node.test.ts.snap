// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-node createSchemaCustomization should call createTypes with custom schema 1`] = `
"
	\\"\\"\\"
	A update or change to a topic, taken from the \`latestChanges\` field in the API reponse for a full topic.

	> Note: this is not a node in its own right so can't be queried directly.

	Don't confuse this with the \`CksChange\` node type.
	\\"\\"\\"
	type CksTopicLatestChange @dontInfer {
		\\"The first day of the month in which this change occurred e.g. \`2016-01-01T00:00:00+00:00\`\\"
		dateFrom: Date! @dateformat

		\\"The last day of the month in which this change occurred e.g. \`2016-01-31T23:59:59+00:00\`\\"
		dateTo: Date! @dateformat

		\\"The type of the change for example *Reviewed*, *Minor update*, *Revised* etc\\"
		title: String!

		\\"A description of the change\\"
		body: String!
	}

	\\"\\"\\"
	A term for a topic, from the \`terms\` field in the API reponse for a full topic.

	> Note: this is not a node in its own right so can't be queried directly.

	For example the topic \\"Achilles tendinopathy\\" might have terms of:
	- Achilles Tendon
	- Tendinopathy
	\\"\\"\\"
	type CksTopicTerm @dontInfer {
		\\"The id of the term e.g. *D001249*\\"
		code: String!

		\\"The term text e.g. *Asthma* or *Tendinopathy*\\"
		term: String!
	}

	\\"\\"\\"
	A full topic object, taken from the */topic/guid* endpoint in the API.
	\\"\\"\\"
	type CksTopic implements Node @dontInfer {
		\\"The lowercase slugified title, used as the path within a URL e.g. *achilles-tendinopathy*\\"
		slug: String!

		\\"The guid id of the topic taken from the feed\\"
		topicId: String!

		\\"The name of the topic e.g. *Achilles tendinopathy*\\"
		topicName: String!

		\\"A summary of the topic in 1 or 2 sentences\\"
		topicSummary: String!

		\\"The latest version number\\"
		number: Int!

		\\"The date of the last revision as text e.g. *Last revised in January 2016*\\"
		lastRevised: String!

		\\"The date by which this topic will be reviewed. Optional as this is null for some topics.\\"
		nextPlannedReviewBy: Date @dateformat

		\\"A list of terms related to this topic. Not every topic has other terms.\\"
		terms: [CksTopicTerm!]!

		\\"A list of alternative names or synonyms\\"
		aliases: [String!]!

		\\"\\"\\"
		A list of specialities to which this topic belongs.

		For example, *Achilles tendinopathy* is in both *Injuries* and *Musculoskeletal*.

		Note: not all topics have specialities defined.
		\\"\\"\\"
		specialities: [CksSpeciality!]! @link(by: \\"name\\")

		\\"The content chapters within the topic, taken from the \`topicHtmlObjects\` field in the API.\\"
		chapters: [CksChapter!]! @link(by: \\"itemId\\")

		\\"The last changes or reviews that the topic has had. Not to be confused with the other node type of \`CksChange\`.\\"
		latestChanges: [CksTopicLatestChange!]!
	}


	\\"\\"\\"
	A clinical speciality that groups topics together

	For example *Injuries* or *Musculoskeletal*.
	\\"\\"\\"
	type CksSpeciality implements Node @dontInfer {
		\\"The name of the speciality e.g. *Injuries*\\"
		name: String!

		\\"The lowercased, slugified title of the speciality used as the path within a URL e.g. *kidney-disease-and-urology*\\"
		slug: String!

		\\"The list of topics that belong to this clinical speciality\\"
		topics: [CksTopic!]! @link(by: \\"topicId\\")
	}


	\\"\\"\\"
	A chapter of content within a topic, taken from the \`topicHtmlObjects\` field in the API.

	Chapters can nest recursively.
	\\"\\"\\"
	type CksChapter implements Node @dontInfer {
		\\"A lowercased slugified chapter title for use in URL paths e.g. *have-i-got-the-right-topic*\\"
		slug: String!

		\\"The guid id of the chapter\\"
		itemId: String!

		\\"The name of the chapter e.g. *Have I got the right topic?*\\"
		fullItemName: String!

		\\"\\"\\"
		The HTML heading for the chapter e.g. \`<h1>Achilles tendinopathy: Summary</h1>\` or \`<h2>Changes</h2>\` or \`<h3>Previous changes</h3>\`.

		Notice two things:
		- the heading tag used corresponds to the \`depth\` of the chapter
		- the text isn't necessarily the same as \`fullItemName\` (although often it is)
		\\"\\"\\"
		htmlHeader: String!

		\\"\\"\\"
		Summary for chapter e.g. 'Covers the management of children and adults with newly diagnosed or suspected asthma.'
		\\"\\"\\"
		summary: String

		\\"The HTML body of the chapter\\"
		htmlStringContent: String!

		\\"\\"\\"
		An identifier for the chapter, used as a CSS class for the containing element in Clarity's Prodigy. e.g. *annualKnowNewEvidence*.

		Note: it's often camel case but sometimes lowercase e.g. *uptodate*.
		\\"\\"\\"
		containerElement: String!

		\\"The 1-based index representing how deeply nested this chapter is. E.g. Summary has a depth of 1.\\"
		depth: Int!

		\\"The 0-based index of the chapter at the current depth\\"
		pos: Int!

		\\"A reference to the topic to which this chapter belongs\\"
		topic: CksTopic! @link(by: \\"topicId\\")

		\\"The direct parent chapter in the nesting. The root chapter has no parent so this field is nullable.\\"
		parentChapter: CksChapter @link(by: \\"itemId\\")

		\\"The chapter at the top of the hierarchy. This property is non-null and references itself for the root chapter.\\"
		rootChapter: CksChapter! @link(by: \\"itemId\\")

		\\"\\"\\"
		Child chapters nested within this one.

		Note: sub chapters are from the the \`children\` field in the API, however,
		we rename them because children is a 'reserved' field in Gatsby
		and is mainly used for transformation relations between nodes
		\\"\\"\\"
		subChapters: [CksChapter!] @link(by: \\"itemId\\")
	}


	\\"\\"\\"
	A monthly change or update to a topic. Taken from the *changes-since* API endpoint.
	\\"\\"\\"
	type CksChange implements Node @dontInfer {
		\\"The type of the change. e.g. *Reviewed* or *New topic* etc. Most changes are *reviewed*.\\"
		title: String!

		\\"A description of the change. Usually 1 or 2 sentences but can be longer\\"
		text: String!

		\\"The topic that was affected by this change\\"
		topic: CksTopic! @link(by: \\"topicId\\")
	}
"
`;
